File: python/raspberrypi/DFRobot_TCS3430.py
""" file DFRobot_TCS3430.py
  # DFRobot_TCS3430 Class infrastructure, implementation of underlying methods
  @copyright   Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
  @licence     The MIT License (MIT)
  @author      [yangfeng]<feng.yang@dfrobot.com> 
  @version  V1.0
  @date  2021-01-26
  @get from https://www.dfrobot.com
  @url https://github.com/DFRobot/DFRobot_TCS3430
"""
import smbus
import time

class DFRobot_TCS3430:
  DFRobot_TCS3430_IIC_ADDR           = 0x39
  
  TCS3430_REG_ENABLE_ADDR            = 0x80
  TCS3430_REG_ATIME_ADDR             = 0x81
  TCS3430_REG_WTIME_ADDR             = 0x83
  TCS3430_REG_AILTL_ADDR             = 0x84
  TCS3430_REG_AILTH_ADDR             = 0x85
  TCS3430_REG_AIHTL_ADDR             = 0x86
  TCS3430_REG_AIHTH_ADDR             = 0x87
  TCS3430_REG_PERS_ADDR              = 0x8C
  TCS3430_REG_CFG0_ADDR              = 0x8D
  TCS3430_REG_CFG1_ADDR              = 0x90
  TCS3430_REG_REVID_ADDR             = 0x91
  TCS3430_REG_ID_ADDR                = 0x92
  TCS3430_REG_STATUS_ADDR            = 0x93
  TCS3430_REG_CH0DATAL_ADDR          = 0x94
  TCS3430_REG_CH0DATAH_ADDR          = 0x95
  TCS3430_REG_CH1DATAL_ADDR          = 0x96
  TCS3430_REG_CH1DATAH_ADDR          = 0x97
  TCS3430_REG_CH2DATAL_ADDR          = 0x98
  TCS3430_REG_CH2DATAH_ADDR          = 0x99
  TCS3430_REG_CH3DATAL_ADDR          = 0x9A
  TCS3430_REG_CH3DATAH_ADDR          = 0x9B
  TCS3430_REG_CFG2_ADDR              = 0x9F
  TCS3430_REG_CFG3_ADDR              = 0xAB
  TCS3430_REG_AZCONFIG_ADDR          = 0xD6
  TCS3430_REG_INTENAB_ADDR           = 0xDD
  
  ENABLEREG_POWER_ON                 = 0x01
  ENABLEREG_POWER_OFF                = 0xFE
  ENABLEREG_ALS_EN                   = 0x02
  ENABLEREG_ALS_DISEN                = 0xFD
  ENABLEREG_WAIT_EN                  = 0x08
  ENABLEREG_WAIT_DISEN               = 0xF7
  
  CONFIG_NO_WLONG                    = 0x80
  CONFIG_WLONG                       = 0x84
  
  CFG1_IR2_EN                        = 0x08
  CFG1_IR2_DISEN                     = 0xF7
  
  CFG2_HIGH_GAIN_EN                  = 0x14
  CFG2_HIGH_GAIN_DISEN               = 0x04
  
  CFG3_INT_READ_CLEAR_EN             = 0x80
  CFG3_INT_READ_CLEAR_DISEN          = 0x10
  CFG3_SAI_EN                        = 0x10
  CFG3_SAI_DISEN                     = 0x80

  AZ_MODE_0                          = 0x7F
  AZ_MODE_1                          = 0x80
  
  ENABLEREG_ALS_INT_EN               = 0x10
  ENABLEREG_ALS_INT_DISEN            = 0x80
  ALS_SATURATION_INTERRUPT_EN        = 0x80
  ALS_SATURATION_INTERRUPT_DISEN     = 0x10
  
  TCS3430_ID                         = 0xDC
  TCS3430_REVISION_ID                = 0x41
  
  ''' 
    @brief  Module init
    @param  bus  Set to IICBus
  '''
  def __init__(self,bus = 1):
    self.__i2cbus = smbus.SMBus(bus)
    self.__i2c_addr = self.DFRobot_TCS3430_IIC_ADDR
    self.__wlong = 0
    self.__atime = 0
    self.__wtime = 0

  ''' 
    @brief  Initialize the device and turn it on
    @return  Whether the device is on or not. True succeed, False failed 
  '''
  def begin(self):
    self.__soft_reset()
    self.__set_power_als_on()
    device_id = self.__get_device_id()
    revision_id =self.__get_revision_id()
    if device_id != self.TCS3430_ID and revision_id != self.TCS3430_REVISION_ID :
      self.__set_device_adc(False)
      self.__set_device_power(False)
      return False
    return True 

  ''' 
    @brief  Config the wait timer 
    @param  mode  set wait-timer,True enable False disenable
  '''
  def set_wait_timer(self,mode=True):
    if mode==True:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR)|self.ENABLEREG_WAIT_EN)
    if mode==False:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR)&self.ENABLEREG_WAIT_DISEN)

  ''' 
    @brief  Set the internal integration time
    @param  atime  the internal integration time(range: 0x00 -0xff)
  '''
  def set_integration_time(self,atime):
    atime = atime & 0xFF
    self.__atime = atime
    self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_ATIME_ADDR, atime)

  ''' 
    @brief  Set wait time 
    @param  wtime  wait time(range: 0x00 -0xff)
  '''
  def set_wait_time(self,wtime):
    wtime = wtime & 0xFF
    self.__wtime = wtime
    self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_WTIME_ADDR, wtime)

  ''' 
    @brief  Set the channel 0 interrupt threshold
    @param  ailt   the low 16 bit values(range: 0x0000 -0xffff)
    @param  aiht   the high 16 bit values(range: 0x0000 -0xffff)
  '''
  def set_interrupt_threshold(self,ailt,aiht):
    ailtl = ailt & 0xFF
    ailth = (ailt>>8) & 0xFF
    aihtl = aiht & 0xFF
    aihth = (aiht>>8) & 0xFF
    self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_AILTL_ADDR, ailtl)
    self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_AILTH_ADDR, ailth)
    self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_AIHTL_ADDR, aihtl)
    self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_AIHTH_ADDR, aihth)

  ''' 
    @brief  Set the channel 0 interrupt Persistence
    @param  apers  Interrupt Persistence(range: 0x00 -0x0f)
  '''
  def set_interrupt_persistence(self,apers):
    apers = apers & 0xFF
    self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_PERS_ADDR, apers)

  ''' 
    @brief  Set the wait long time
    @param  mode  True enable  False disenable
  '''
  def set_wait_long_time(self,mode=True):
    if mode == True:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG0_ADDR, self.CONFIG_WLONG)
      self.__wlong = 1
    if mode == False:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG0_ADDR, self.CONFIG_NO_WLONG)
      self.__wlong = 0

  ''' 
    @brief  Set the ALS gain 
    @param  gain  the value of gain(range: 0x00 -0x03)
  '''
  def set_als_gain(self,gain):
    gain = gain & 0xFF
    self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG1_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG1_ADDR)|gain)

  ''' 
    @brief get channel 0 value
    @return  the z data
  '''
  def get_z_data(self):
    vlaue = self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CH0DATAL_ADDR)
    data = vlaue | (self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CH0DATAH_ADDR)<<8)
    return data 

  ''' 
    @brief get channel 1 value
    @return  the y data
  '''
  def get_y_data(self):
    value = self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CH1DATAL_ADDR)
    value = value | (self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CH1DATAH_ADDR)<<8)
    return value 

  ''' 
    @brief get channel 2 value
    @return  the IR1 data
  '''
  def get_ir1_data(self):
    value = self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CH2DATAL_ADDR)
    value = value | (self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CH2DATAH_ADDR)<<8)
    return value 
    
  ''' 
    @brief get channel 3 value
    @return  the X data
  '''
  def get_x_data(self):
    value = self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CH3DATAL_ADDR)
    value = value | (self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CH3DATAH_ADDR)<<8)
    return value 

  ''' 
    @brief get channel 3 value
    @return  the IR2 data
  '''
  def get_ir2_data(self):
    self.__set_ir2_channel(True)
    if (self.__wlong):
      delaytime = ((self.__atime+1)*2.78 + (self.__wtime+1)*33.4)/1000
    else:
      delaytime =((self.__atime+1)*2.78 + (self.__wtime+1)*2.78)/1000
    time.sleep(delaytime)
    value = self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CH3DATAL_ADDR)
    value = value | (self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CH3DATAH_ADDR)<<8)
    self.__set_ir2_channel(False)
    time.sleep(delaytime)
    return value 

  ''' 
    @brief  Set the ALS  128x gain 
    @param  mode  True enable  False disenable
  '''
  def set_als_high_gain(self,mode=True):
    if mode == True:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG2_ADDR, self.CFG2_HIGH_GAIN_EN)
    if mode == False:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG2_ADDR, self.CFG2_HIGH_GAIN_DISEN)

  '''
    @brief  If this bit is set, all flag bits in the STATUS register will be reset whenever the STATUS register is read over I2C. 
    @param  mode  True enable  False disenable
  '''
  def set_int_read_clear(self,mode=True):
    if mode == True:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG3_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG3_ADDR)|self.CFG3_INT_READ_CLEAR_EN)
    if mode == False:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG3_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG3_ADDR)|self.CFG3_INT_READ_CLEAR_DISEN)

  '''
    @brief  Config the function of 'sleep after interruption'
    @param  mode  True enable  False disenable
  '''
  def set_sleep_after_interrupt(self,mode=True):
    if mode == True:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG3_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG3_ADDR)|self.CFG3_SAI_EN)
    if mode == False:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG3_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG3_ADDR)|self.CFG3_SAI_DISEN)

  '''
    @brief  set aotuzero mode
    @param  mode  0,Always start at zero when searching the best offset value  1,Always start at the previous (offset_c) with the auto-zero mechanism
  '''
  def set_auto_zero_mode(self,mode=0):
    if(mode==1):
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_AZCONFIG_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_AZCONFIG_ADDR)|self.AZ_MODE_1)
    if(mode==0):
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_AZCONFIG_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_AZCONFIG_ADDR)&self.AZ_MODE_0)

  ''' 
    @brief  set autozero automatically every nth ALS iteration)
    @param  iteration_type  0,never  7F,only at first ALS cycle  n, every nth time
  '''
  def set_auto_zero_nth_iteration(self,iteration_type):
    iteration_type = iteration_type & 0x7F
    self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_AZCONFIG_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_AZCONFIG_ADDR)|iteration_type)

  '''
    @brief  Config the ambient light sensing interruption
    @param  mode  True enable  False disenable
  '''
  def set_als_interrupt(self,mode=True):
    self.set_int_read_clear(True);
    if(mode==True):
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_INTENAB_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_INTENAB_ADDR)|self.ENABLEREG_ALS_INT_EN)
    if(mode==False):
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_INTENAB_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR)&self.ENABLEREG_ALS_INT_DISEN)

  '''
    @brief  Config the ALS saturation interription
    @param  mode  True enable  False disenable
  '''
  def set_als_saturation_interrupt(self,mode=True):
    self.set_int_read_clear(True);
    if(mode==True):
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_INTENAB_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_INTENAB_ADDR)|self.ALS_SATURATION_INTERRUPT_EN)
    if(mode==False):
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_INTENAB_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_INTENAB_ADDR)&self.ALS_SATURATION_INTERRUPT_DISEN)

  '''
    @brief  Get the status of the device
  '''
  def get_device_status(self):
    self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_STATUS_ADDR)

  '''
    @brief  Access to IR channel; allows mapping of IR channel on channel 3.
    @param  mode  True enable  False disenable
  '''
  def __set_ir2_channel(self,mode=True):
    if mode == True:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG1_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG1_ADDR)|self.CFG1_IR2_EN)
    if mode == False:
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG1_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_CFG1_ADDR)&self.CFG1_IR2_DISEN)

  '''
    @brief  Activating the internal oscillator to permit the timers and ADC channels to operate ,and activing the ALS function
  '''
  def __set_power_als_on(self):

    self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR, self.ENABLEREG_ALS_EN|self.ENABLEREG_POWER_ON)

  '''
    @brief  Activating the internal oscillator to permit the timers and ADC channels to operate
    @param  mode  True enable  False disenable
  '''
  def __set_device_power(self,mode=True):
    if(mode==True):
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR)|self.ENABLEREG_POWER_ON)
    if(mode==False):
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR)&self.ENABLEREG_POWER_OFF)

  '''
    @brief  Activating the four-channel ADC
    @param  mode  True enable  False disenable
  '''
  def __set_device_adc(self,mode=True):
    if(mode==True):
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR)|self.ENABLEREG_ALS_EN)
    if(mode==False):
      self.__i2cbus.write_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR, self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_ENABLE_ADDR)&self.ENABLEREG_ALS_DISEN)

  '''
    @brief  get the revision id
    @return  the revision id
  '''
  def __get_revision_id(self):
    return self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_REVID_ADDR)

  '''
    @brief  get the device id
    @return  the device id
  '''
  def __get_device_id(self):
    a = self.__i2cbus.read_byte_data(self.__i2c_addr,self.TCS3430_REG_ID_ADDR)
    return a

  '''
    @brief  Initializes all registers of the device
  '''
  def __soft_reset(self):
    self.set_wait_timer(False)
    self.set_integration_time(0x23)
    self.set_wait_time(0)
    self.set_wait_long_time(False)
    self.set_als_gain(3)
    self.set_als_high_gain(False)
    self.set_int_read_clear(False)
    self.set_sleep_after_interrupt(False)
    self.set_auto_zero_mode(0)
    self.set_auto_zero_nth_iteration(0x7F)
    self.set_als_saturation_interrupt(False)
    self.set_als_interrupt(False)

    

File: DFRobot_TCS3430.cpp
/*!
 * @file DFRobot_TCS3430.cpp
 * @brief Implementation of DFRobot_TCS3430 class
 * @copyright Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
 * @SKU SEN0403
 * @licence The MIT License (MIT)
 * @author [yangfeng]<feng.yang@dfrobot.com>
 * @version V1.0
 * @date 2021-01-26
 * @url  https://github.com/DFRobot/DFRobot_TCS3400
 */
#include <DFRobot_TCS3430.h>
DFRobot_TCS3430::DFRobot_TCS3430(TwoWire *pWire):
_pWire(pWire),_deviceAddr(DFRobot_TCS3430_ICC_ADDR)
{
  _atime = 0;
  _wtime = 0;
  _wlong = 0;
  _enableReg.pon = 0;
  _enableReg.aen = 0;
  _enableReg.reservedBit2 = 0;
  _enableReg.wen = 0;
  _enableReg.reservedBit4_7 = 0;
  
  _cfg1Reg.again = 0;
  _cfg1Reg.reservedBit2 = 0;
  _cfg1Reg.amux = 0;
  _cfg1Reg.reservedBit4_7 = 0;
  
  _cfg3Reg.reservedBit0_3 = 0;
  _cfg3Reg.sai = 0;
  _cfg3Reg.reservedBit5_6 = 0;
  _cfg3Reg.intReadClear = 0;
  
  _AZCfgReg.azNTHIteration = 0x7F;
  _AZCfgReg.azMode = 0;
  
  _intEnabReg.reservedBit0_3 = 0;
  _intEnabReg.aien = 0;
  _intEnabReg.reservedBit5_6 =0;
  _intEnabReg.asien = 0;
}

bool DFRobot_TCS3430::begin()
{
  _pWire->begin();
  softReset();
  setPowerALSADC();
  if((getDeviceID()!=TCS3430_ID) || (getRevisionID() != TCS3430_REVISION_ID)){
    disableALSADC();
    powerOFF();
    return false;
  }
  return true;
}

void DFRobot_TCS3430::setIntegrationTime(uint8_t aTime)
{
  _atime = aTime;
  write(eRegATIMEAddr,aTime);
}

void DFRobot_TCS3430::setWaitTime(uint8_t wTime)
{
  _wtime = wTime;
  write(eRegWTIMEAddr,wTime);
}

void DFRobot_TCS3430:: setPowerALSADC()
{
  _enableReg.pon = 1;
  _enableReg.aen = 1;
  write(eRegENABLEAddr,*((uint8_t*)(&_enableReg)));
}

void DFRobot_TCS3430:: powerOFF()
{
  _enableReg.pon = 0;
  write(eRegENABLEAddr,*((uint8_t*)(&_enableReg)));
}

void DFRobot_TCS3430:: disableALSADC()
{
  _enableReg.aen = 0;
  write(eRegENABLEAddr,*((uint8_t*)(&_enableReg)));
}

void DFRobot_TCS3430:: setWaitTimer(bool mode)
{
  if(mode = true){
    _enableReg.wen = 1;
    write(eRegENABLEAddr,*((uint8_t*)(&_enableReg)));
  }
  else{
    _enableReg.wen = 0;
    write(eRegENABLEAddr,*((uint8_t*)(&_enableReg)));
  }
}

void DFRobot_TCS3430:: setInterruptPersistence(uint8_t apers)
{
  write(eRegPERSAddr,apers);
}

void DFRobot_TCS3430:: setWaitLong(bool mode)
{
  if(mode){
    write(eRegCFG0Addr,DFRobot_TCS3430_CONFIG_WLONG);
    _wlong =1;

  }else{
    write(eRegCFG0Addr,DFRobot_TCS3430_CONFIG_NO_WLONG);
    _wlong =0;
  }
}


void DFRobot_TCS3430:: setALSGain(uint8_t aGain)
{
  _cfg1Reg.again=aGain;
  write(eRegCFG1Addr,*((uint8_t*)(&_cfg1Reg)));
}

void DFRobot_TCS3430:: setIR2Channel(bool mode)
{
  if(mode){
    _cfg1Reg.amux=1;
    write(eRegCFG1Addr,*((uint8_t*)(&_cfg1Reg)));
  }else{
    _cfg1Reg.amux=0;
    write(eRegCFG1Addr,*((uint8_t*)(&_cfg1Reg)));
  }
}


uint8_t DFRobot_TCS3430:: getRevisionID()
{
  return uint8_t(read(eRegREVIDAddr,1));
}

uint8_t DFRobot_TCS3430:: getDeviceID()
{
  return uint8_t(read(eRegIDAddr,1));
}

uint8_t DFRobot_TCS3430:: getDeviceStatus()
{
  return uint8_t(read(eRegSTATUSAddr,1));
}

uint16_t DFRobot_TCS3430:: getZData()
{
  return read(eRegCH0DATALAddr,TWO_BYTE);
}

uint16_t DFRobot_TCS3430:: getYData()
{
  return DFRobot_TCS3430:: read(eRegCH1DATALAddr,TWO_BYTE);
}

uint16_t DFRobot_TCS3430:: getIR1Data()
{
  return read(eRegCH2DATALAddr,TWO_BYTE);
}

uint16_t DFRobot_TCS3430:: getXData()
{
  return read(eRegCH3DATALAddr,TWO_BYTE);
}

uint16_t DFRobot_TCS3430:: getIR2Data()
{
  disableALSADC();
  setIR2Channel(true);
  _enableReg.aen = 1;
  write(eRegENABLEAddr,*((uint8_t*)(&_enableReg)));
  uint16_t delayTime = 0;
  if(_wlong){
    delayTime = (_atime+1)*2.78 + (_wtime+1)*33.4;
  }else{
    delayTime = (_atime+1)*2.78 + (_wtime+1)*2.78; 
  }
  delay(delayTime);
  uint16_t value = read(eRegCH3DATALAddr,TWO_BYTE);
  disableALSADC();
  setIR2Channel(false);
  _enableReg.aen = 1;
  write(eRegENABLEAddr,*((uint8_t*)(&_enableReg)));
  delay(delayTime);
  return value;
}

void DFRobot_TCS3430:: setHighGAIN(bool mode)
{
  if (mode){
    write(eRegCFG2Addr,DFRobot_TCS3430_HGAIN_ENABLE);
  }else{
    write(eRegCFG2Addr,DFRobot_TCS3430_HGAIN_DISABLE);
  }
}

void DFRobot_TCS3430:: setIntReadClear(bool mode)
{
  if (mode){
    _cfg3Reg.intReadClear = 1;
    write(eRegCFG3Addr,*((uint8_t*)(&_cfg3Reg)));
  }else{
    _cfg3Reg.intReadClear = 0;
    write(eRegCFG3Addr,*((uint8_t*)(&_cfg3Reg)));
  }
}

void DFRobot_TCS3430:: setSleepAfterInterrupt(bool mode )
{

  if (mode){
    _cfg3Reg.sai = 1;
    write(eRegCFG3Addr,*((uint8_t*)(&_cfg3Reg)));
  }else{
    _cfg3Reg.sai = 0;
    write(eRegCFG3Addr,*((uint8_t*)(&_cfg3Reg)));
  }
}

void DFRobot_TCS3430:: setAutoZeroMode(uint8_t mode)
{
  if(mode == 0){
    _AZCfgReg.azMode = 0;
  }
  if(mode == 1){
    _AZCfgReg.azMode = 1;
  }
  write(eRegAZCONFIGAddr,*((uint8_t*)(&_AZCfgReg)));
}

void DFRobot_TCS3430:: setAutoZeroNTHIteration(uint8_t value)
{
  _AZCfgReg.azNTHIteration = value &0x7F;
  write(eRegAZCONFIGAddr,*((uint8_t*)(&_AZCfgReg)));
}

void DFRobot_TCS3430:: setALSSaturationInterrupt(bool mode )
{
  setIntReadClear(true);
  if(mode){
    _intEnabReg.asien = 1;
    write(eRegINTENABAddr,*((uint8_t*)(&_intEnabReg)));
  }else{
    _intEnabReg.asien = 0;
    write(eRegINTENABAddr,*((uint8_t*)(&_intEnabReg)));
  }
}

void DFRobot_TCS3430:: setALSInterrupt(bool mode )
{
  setIntReadClear(true);
  if(mode){
    _intEnabReg.aien = 1;
    write(eRegINTENABAddr,*((uint8_t*)(&_intEnabReg)));
  }else{
    _intEnabReg.aien = 0;
    write(eRegINTENABAddr,*((uint8_t*)(&_intEnabReg)));
  }
}


void DFRobot_TCS3430:: setCH0IntThreshold(uint16_t thresholdL,uint16_t thresholdH)
{
  uint8_t ailtl = thresholdL&0x00FF;
  uint8_t ailth = (thresholdL&0xFF00)>>8;
  uint8_t aihtl = thresholdH&0x00FF;
  uint8_t aihth = (thresholdH&0xFF00)>>8;
  write(eRegAILTLAddr,ailtl);
  write(eRegAILTHAddr,ailth);
  write(eRegAIHTLAddr,aihtl);
  write(eRegAIHTHAddr,aihth);
}

void DFRobot_TCS3430:: softReset()
{
  setWaitTimer(false);
  setIntegrationTime(0x23);
  setWaitTime(0);
  setWaitLong(false);
  setALSGain(3);
  setHighGAIN(false);
  setIntReadClear(false);
  setSleepAfterInterrupt(false);
  setAutoZeroMode(0);
  setAutoZeroNTHIteration(0x7f);
  setALSInterrupt(false);
  setALSSaturationInterrupt(false);
}

void DFRobot_TCS3430:: write(uint8_t regAddr,uint8_t value)
{
  _pWire->beginTransmission(_deviceAddr);
  _pWire->write(regAddr);
  _pWire->write(value);
  _pWire->endTransmission();
}

uint16_t DFRobot_TCS3430:: read(uint8_t regAddr,uint8_t readNum)
{
  uint16_t value=0;
  _pWire->beginTransmission(_deviceAddr);
  _pWire->write(regAddr);
  _pWire->endTransmission();
  _pWire->requestFrom(_deviceAddr, readNum);
  if(readNum==1){
    value = _pWire->read();
  }else if(readNum == 2){
    value = _pWire->read();
    value |= _pWire->read()<<8;
  }
  return value;
}
